drivers/vulkan/rendering_device_driver_vulkan.cpp

```diff
@@ -1353,6 +1391,34 @@ void RenderingDeviceDriverVulkan::buffer_unmap(BufferID p_buffer) {
 	vmaUnmapMemory(allocator, buf_info->allocation.handle);
 }
 
+void RenderingDeviceDriverVulkan::wait_for_present(SwapChainID p_swap_chain) {
+	if (!present_wait_support) {
+		return;
+	}
+
+	SwapChain *swap_chain = (SwapChain *)(p_swap_chain.id);
+	if (!swap_chain->vk_swapchain || !swap_chain->needs_wait) {
+		return;
+	}
+
+	int latency = swap_chain->images.size() - 1;
+	if (latency < 0 || swap_chain->present_id <= static_cast<unsigned>(latency)) {
+		return;
+	}
+
+	uint64_t wait_present_id = swap_chain->present_id - latency;
```

Comment:
Should we be subtracting `latency` here? Shouldn't we just use the present id as given? The documentation for [vkWaitForPresentKHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkWaitForPresentKHR.html) says:

```
vkWaitForPresentKHR waits for the presentId associated with swapchain to be increased in value so that it is at least equal to presentId.
```

So if `wait_present_id` is too low, presumably vkWaitForPresentKHR won't perform the wait, thus we won't see an improvement in latency.

Reply:
If it uses the latest present id then it will only ever be queueing one present image in the swap chain. I may be wrong here but my understanding is that, when `swapchain_image_count` is set to, say 3, one is always held as the front buffer, the remaining 2 images can be acquired, drawn onto and queued for present. We want to queue two images, but only wait until the first in queue is presented.

Though this made me realize the code actually need to subtract one from the latency to get the desired effect.